bee-message/src/payload/milestone/essence.rs

```diff
@@ -129,62 +163,44 @@ impl MilestoneEssence {
-        self.index.pack(writer)?;
-        self.timestamp.pack(writer)?;
-        self.parents.pack(writer)?;
-        self.merkle_proof.pack(writer)?;
-        self.next_pow_score.pack(writer)?;
-        self.next_pow_score_milestone_index.pack(writer)?;
-        (self.public_keys.len() as u8).pack(writer)?;
-        for public_key in &self.public_keys {
-            public_key.pack(writer)?;
-        }
-        option_payload_pack(writer, self.receipt.as_ref())?;
+    type UnpackError = Error;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.index.pack(packer)?;
+        self.timestamp.pack(packer)?;
+        self.parents.pack(packer)?;
+        self.merkle_proof.pack(packer)?;
+        self.next_pow_score.pack(packer)?;
+        self.next_pow_score_milestone_index.pack(packer)?;
+        self.public_keys.pack(packer)?;
+        self.receipt.pack(packer)?;
 
         Ok(())
     }
 
-    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error> {
-        let index = MilestoneIndex::unpack_inner::<R, CHECK>(reader)?;
-        let timestamp = u64::unpack_inner::<R, CHECK>(reader)?;
-        let parents = Parents::unpack_inner::<R, CHECK>(reader)?;
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let index = MilestoneIndex::unpack::<_, VERIFY>(unpacker).infallible()?;
+        let timestamp = u64::unpack::<_, VERIFY>(unpacker).infallible()?;
+        let parents = Parents::unpack::<_, VERIFY>(unpacker)?;
 
-        let merkle_proof = <[u8; MilestoneEssence::MERKLE_PROOF_LENGTH]>::unpack_inner::<R, CHECK>(reader)?;
+        let merkle_proof = <[u8; MilestoneEssence::MERKLE_PROOF_LENGTH]>::unpack::<_, VERIFY>(unpacker).infallible()?;
 
-        let next_pow_score = u32::unpack_inner::<R, CHECK>(reader)?;
-        let next_pow_score_milestone_index = u32::unpack_inner::<R, CHECK>(reader)?;
+        let next_pow_score = u32::unpack::<_, VERIFY>(unpacker).infallible()?;
+        let next_pow_score_milestone_index = u32::unpack::<_, VERIFY>(unpacker).infallible()?;
 
-        let public_keys_len = u8::unpack_inner::<R, CHECK>(reader)? as usize;
-        let mut public_keys = Vec::with_capacity(public_keys_len);
-        for _ in 0..public_keys_len {
-            public_keys.push(<[u8; MilestoneEssence::PUBLIC_KEY_LENGTH]>::unpack_inner::<R, CHECK>(
-                reader,
-            )?);
-        }
+        let public_keys = VecPrefix::<[u8; Self::PUBLIC_KEY_LENGTH], PublicKeyCount>::unpack::<_, VERIFY>(unpacker)
+            .map_packable_err(|err| Error::MilestoneInvalidSignatureCount(err.into_prefix().into()))?;
 
-        let (_, receipt) = option_payload_unpack::<R, CHECK>(reader)?;
+        let receipt = OptionalPayload::unpack::<_, VERIFY>(unpacker)?;
 
-        Self::new(
+        Self::from_vec_prefix(
```

Comment:
Good ol' verify_with ?

Reply:
sadly no :/ one of the checks uses both `next_pow_score` and `next_pow_score_milestone_index`

Reply:
sad reacts only

Reply:
I won't say you that I could change 

```rust
pub struct MilestoneEssence {
    index: MilestoneIndex,
    timestamp: u64,
    parents: Parents,
    merkle_proof: [u8; MilestoneEssence::MERKLE_PROOF_LENGTH],
    next_pow_score: u32,
    next_pow_score_milestone_index: u32,
    public_keys: VecPrefix<[u8; MilestoneEssence::PUBLIC_KEY_LENGTH], PublicKeyCount>,
    receipt: OptionalPayload,
}
```
by
```rust
pub struct MilestoneEssence {
    index: MilestoneIndex,
    timestamp: u64,
    parents: Parents,
    merkle_proof: [u8; MilestoneEssence::MERKLE_PROOF_LENGTH],
    next_pow_score_and_milestone_index: (u32, u32),
    public_keys: VecPrefix<[u8; MilestoneEssence::PUBLIC_KEY_LENGTH], PublicKeyCount>,
    receipt: OptionalPayload,
}
```
because tuples don't implement packable yet.

Reply:
I wouldn't change the layout just for the sake of derivation anyway

Reply:
technically the layout is unchanged :trollface: nobody can access these fields directly, only via methods.

Reply:
Interesting :trollface: